haystack/document_stores/memory.py

```diff
@@ -64,6 +65,8 @@ def __init__(
                                    you have at least `embedding_dim`*`scoring_batch_size`*4 bytes available in GPU memory.
                                    Since the data is originally stored in CPU memory there is little risk of overruning memory
                                    when running on CPU.
+        :param devices: List of GPU devices to limit inference to certain GPUs and not use all available ones (e.g. [torch.device('cuda:0')]).
```

Comment:
I would consider changing the doc string to something like `List of torch devices (e.g. cuda, cpu, mps) to limit inference to certain devices ...` since a `torch.device` does not have to be a GPU. What do you think?

Reply:
Oh sorry, I just read the docstring for `initialize_device_settings` and see that `devices` is only used if `use_cuda` is True so I think we can leave this unchanged. Perhaps add a sentence explaining that this list is only used if `use_gpu` is set to `True`

Reply:
+1, excellent idea

Reply:
I tried this one but it doesn't seem like it is going to work out of the box. For Tutorial 3, during pipeline run I get the following exception:
```
  File "/Users/vblagoje/workspace/haystack/haystack/pipelines/base.py", line 512, in run
    node_output, stream_id = self._run_node(node_id, node_input)
  File "/Users/vblagoje/workspace/haystack/haystack/pipelines/base.py", line 439, in _run_node
    return self.graph.nodes[node_id]["component"]._dispatch_run(**node_input)
  File "/Users/vblagoje/workspace/haystack/haystack/nodes/base.py", line 201, in _dispatch_run
    return self._dispatch_run_general(self.run, **kwargs)
  File "/Users/vblagoje/workspace/haystack/haystack/nodes/base.py", line 245, in _dispatch_run_general
    output, stream = run_method(**run_inputs, **run_params)
  File "/Users/vblagoje/workspace/haystack/haystack/nodes/reader/base.py", line 84, in run
    results = predict(query=query, documents=documents, top_k=top_k)
  File "/Users/vblagoje/workspace/haystack/haystack/nodes/reader/base.py", line 145, in wrapper
    ret = fn(*args, **kwargs)
  File "/Users/vblagoje/workspace/haystack/haystack/nodes/reader/farm.py", line 867, in predict
    predictions = self.inferencer.inference_from_objects(
  File "/Users/vblagoje/workspace/haystack/haystack/modeling/infer.py", line 596, in inference_from_objects
    return self.inference_from_dicts(
  File "/Users/vblagoje/workspace/haystack/haystack/modeling/infer.py", line 578, in inference_from_dicts
    return Inferencer.inference_from_dicts(
  File "/Users/vblagoje/workspace/haystack/haystack/modeling/infer.py", line 349, in inference_from_dicts
    return list(predictions)
  File "/Users/vblagoje/workspace/haystack/haystack/modeling/infer.py", line 424, in _inference_with_multiprocessing
    predictions = self._get_predictions_and_aggregate(dataset, tensor_names, baskets)
  File "/Users/vblagoje/workspace/haystack/haystack/modeling/infer.py", line 512, in _get_predictions_and_aggregate
    logits = self.model.forward(
  File "/Users/vblagoje/workspace/haystack/haystack/modeling/model/adaptive_model.py", line 484, in forward
    output_tuple = self.language_model.forward(
  File "/Users/vblagoje/workspace/haystack/haystack/modeling/model/language_model.py", line 364, in forward
    return self.model(**params, return_dict=return_dict)
  File "/Users/vblagoje/miniconda3/envs/haystack/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/vblagoje/miniconda3/envs/haystack/lib/python3.8/site-packages/transformers/models/roberta/modeling_roberta.py", line 841, in forward
    embedding_output = self.embeddings(
  File "/Users/vblagoje/miniconda3/envs/haystack/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/vblagoje/miniconda3/envs/haystack/lib/python3.8/site-packages/transformers/models/roberta/modeling_roberta.py", line 128, in forward
    inputs_embeds = self.word_embeddings(input_ids)
  File "/Users/vblagoje/miniconda3/envs/haystack/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/vblagoje/miniconda3/envs/haystack/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 160, in forward
    return F.embedding(
  File "/Users/vblagoje/miniconda3/envs/haystack/lib/python3.8/site-packages/torch/nn/functional.py", line 2206, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: Placeholder storage has not been allocated on MPS device!
```

And yes, I have initialised FarmReader with:
```
reader = FARMReader(model_name_or_path="deepset/roberta-base-squad2", devices=[torch.device("mps")])
```

Do you @sjrl have M1/M2 silicone to try it out as well? Use `vblagoje:t_fix_devices` branch, of course

Reply:
Yes, I saw that one too, but I have the latest build from a few days ago. I traced this issue on QA [inferencer](https://github.com/deepset-ai/haystack/blob/main/haystack/modeling/infer.py#L229) class constructor not passing devices parameter. But even with this fix my kernel dies due to some MPS torch unsupported ops. And the fallback to cpu they suggest doesn't work (the kernel dies). But using `TransformerReader` works :-)

Reply:
I'll add the `devices` parameter in infer.py but we'll have to cleanup it up soon after we integrate this PR. The multi-processing part should be eviscerated asap.

Reply:
> I traced this issue to QA [inferencer](https://github.com/deepset-ai/haystack/blob/main/haystack/modeling/infer.py#L229) class constructor not passing devices parameter.

Nice, great catch!

> I'll add the devices parameter in infer.py but we'll have to cleanup it up soon after we integrate this PR. 

What cleanup do you mean exactly here?

> But even with this fix my kernel dies due to some MPS torch unsupported ops.

Does this mean we need to wait on PyTorch to support more ops with MPS before we can use a `FARMReader` on MPS?

Reply:
We can like for this [comment](https://github.com/pytorch/pytorch/issues/77764#issuecomment-1131413007)

Reply:
> Using CPU as fallback blew up the kernel. Seems risky even if it worked. So it seems like we'll have to wait a bit more.

It seems like the fallback only works if you build pytorch from source https://github.com/pytorch/pytorch/issues/77764#issuecomment-1133035881

Reply:
There is also a separate issue for `aten::cumsum.out` in PyTorch here: https://github.com/pytorch/pytorch/issues/79112